### ISSUE TITLE
Web Application Firewall Integration

### ISSUE THREAD

**Issue by blazinaj** [aws-amplify/amplify-hosting#36]
**Is your feature request related to a problem? Please describe.**
We are trying to use an Amplify-CLI / Amplify-Console powered app in Production. A business (security) requirement is that we use a Web Application Firewall. Currently there is no way to implement that with the Amplify Console, so we will have to host our production app using a different service. 

See: #8 

We are a little sad about this, as the Amplify Console has 90% of the functionality needed for hosting a production level app, but "Only Basic Auth security" is a major deal breaker for us.

**Describe the solution you'd like**
Allow us to create our own CloudFront distribution, with an associated WAF, and direct the Amplify Console Hosting through that domain, instead of the automatically generated "always public" domain (e.g. `production.abcxyz.amplifyapp.com`)

**Comment by garyleefight** [comment]
We have cut a feature request in our backlog. Thanks for your feedback

**Comment by swaminator** [comment]
@blazinaj: what if we were to implement an IP whitelisting feature? Would that be sufficient vs WAF support?

**Comment by blazinaj** [comment]
> @blazinaj: what if we were to implement an IP whitelisting feature? Would that be sufficient vs WAF support?

Not quite sufficient for our use. We won't be able to fully restrict access based on IP address, but we would like to incorporate WAF rules like restricting access to only the United States. 

Other WAF features like pattern matching and logging are also important to us.

Following your workaround and creating a custom CloudFront distribution (with WAF), and using the amplifyapp url as the Origin seems like a legitimate (and fairly simple to set up) way to incorporate a firewall, but we would need to be able to configure the original domain.

I may be wrong, but I believe that having a firewall in place is a Must for many applications, even if the URL needs to be publicly accessible.

**Comment by danieladams456** [comment]
We would like to use AWS Amplify to host apps for our internal users.  For that use case, IP whitelist would be sufficient.

**Comment by connor-rw** [comment]
Just wanted to add that using a WAF is a requirement for my team as well, so this would be a huge feature / value-add for us.

Thanks for all of your great work!

**Comment by neilbts** [comment]
WAF would be fantastic but IP whitelisting would be sufficient. Thanks!

**Comment by rkaplan** [comment]
+1 for IP whitelisting, this would be extremely helpful

**Comment by swaminator** [comment]
@rkaplan @neilbts @connor-rw @danieladams456 @blazinaj thanks for all the feedback. So far we've received requirements along a spectrum:
1. IP Whitelisting: Ability to restrict access based on IP addresses.
2. WAF: Ability to setup a firewall to use features such as geographic restriction, pattern matching.
3. VPC: While we haven't heard it here, we've had customers ask us to offer a managed hosting in their internal VPC.

We will update you when we make progress on these!

**Comment by danieladams456** [comment]
@swaminator VPC would be awesome as we could then easily put it behind ALB OIDC auth.  That would simplify things over having static on Amplify, which then has to trap denies/login redirects on ajax calls and do the appropriate action to kick off the login flow.

It also would mostly do away with the need for #184

**Comment by LechuckThePirate** [comment]
Yes ... in my company we definitely need VPC integration, since our webportal is not accesible from the Internet, but from a VPN connection from clients

**Comment by abhishekdixit1508** [comment]
@swaminator Thanks for taking this up to the next level. Just checking if you have any updates on the implementation? Regards Abhi

**Comment by teemuniiranen** [comment]
I thought that this could be done by just specifying WebACLId to the generated CloudFormation template under CloudFront distribution but seems like it is not used when requested through amplifyapp.com URLs (although it was created correctly).

**Comment by IsaacTrevino** [comment]
+1 This feature would definitely help our company's web app allowing certain networks gain access like covid-19 work from home situation, and a automated WAF since I am quite new to firewalls.

**Comment by toritroniks** [comment]
+1 I also would love to have ip whitelisting.

**Comment by rrsai** [comment]
Just to add to the use case here:

Whenever I develop an app, I initially don't want to fully expose it to the world, but I do want to share it directly with stakeholders so that they can see the work and progress.  So I need a WAF from day 1 (or, suboptimally, ip whitelisting).

Cloudfront sites currently allow for using WAF (even if I had to manually incorporate it, but amplify hosting does not appear to).

**Comment by jabrennem** [comment]
I also have a similar situation where I could heavily use Amplify for many projects at work for quick deployments, but unfortunately cannot use it because they are for an internal facing audience as well. Is there any progress on this feature?

**Comment by tchalvak** [comment]
One approach to get a waf in play is to set up hosting with the cloudfront & s3 option and then attach a waf to that cloudfront.

**Comment by denonade** [comment]
> One approach I found to get a waf in play is to set up hosting with the cloudfront & s3 option and then attach a waf to that cloudfront.
@tchalvak, Is it feasible that pointing the public domain name of Amplify app in Cloudfront as orgin? If so,  is the traffic pass through internet back and forth to Amplify ?

**Comment by tchalvak** [comment]
As long as you pick the non-amplify, cloudfront hosting, you can attach a waf to that cloudfront.  If you use the amplify as an origin for a cloudfront, I haven’t tried that, and would expect it to get very annoying because I expect amplify hosting is an invisible cloudfront of theirs. With amplify hosting you do not have a bucket under your control.

So then you would be pointing a cloudfront origin at cloudfront edges, I imagine the caching would get very hard to manage.  But easy to try it, I suppose.

**Comment by blazinaj** [comment]
>As long as you pick the non-amplify, cloudfront hosting, you can attach a waf to that cloudfront.

Yeah, we currently are hosting using the Amplify CLI Hosting Category with Cloudfront and attached our WAF and it works fine. We then just manually cloned the Amplify Console CI functionality using AWS Code Pipeline and stopped using Amplify Console.

> If you use the amplify as an origin for a cloudfront

You can use the Amplify Console CF as your own Cloudfront Origin, but since you can't configure the Amplify Console managed Cloudfront distribution, there is no way to disable it. Traffic can potentially bypass your Cloudfront+WAF and go through the other one if they can find the domain

**Comment by maczamora** [comment]
Sounds like there is some work arounds that have been attempted using Cloudfront... We are needing to figure out a way to allow a source IP Address to our front-end and validate it's the white listed IP Address if not return to the previous service. Sounds like there isn't a direct way to do this. I'm thinking at this point to have that functionality I'd probably have to migrate away from Amplify and host the front-end on either a container or ec2 instance? Correct me if I'm wrong here.

**Comment by denlcy** [comment]
Almost half year gone, I wanna check up is the existing Amplify still miss the WAF, IP whitelisting and VPC private endpoint integration?

**Comment by ckho-wkcda** [comment]
Is it still on roadmap?

**Comment by swaminator** [comment]
@blazinaj would you be willing to chat with us on requirements? We are (finally) picking this up.

**Comment by swaminator** [comment]
@ckho-wkcda same. Would you be willing to talk to us about the feature?

**Comment by denonade** [comment]
@swaminator , below are two high level requirements, please check, thanks a lot.

1) For the CloudFront embedded in Amplify implicitly, is it support WAF, IP whitelisting those ingress control features?
2) If the use case is not public-facing, can Amplify integrate with VPN endpoint or private link of API gateway such that the Amplify can be a sub-system from the central system hosted in the VPC landscape.

**Comment by swaminator** [comment]
@denonade can you elaborate on your 2nd usecase?

**Comment by denonade** [comment]
![image](https://user-images.githubusercontent.com/14343947/117747262-65c51d00-b240-11eb-8bc4-c447993fa5ec.png)
![image](https://user-images.githubusercontent.com/14343947/117747992-cb65d900-b241-11eb-97a5-eb1bd15a792d.png)

Nano URL service
1. End user login and register a long url from the web app developed and hosted by Amplify
2. End user got the nano url from the web app, and share it in the SNS. Other end user browse the tiny url

From the architecture diagram above, you can see that
1. In order to have ingress control, additional Clouldfront with WAF need to placed in front of Amplify
2. As the VPC and private domain name are not supported by AWS Amplify at this 
moment, private API is not applicable to the REST API calls (POST /stage/resPool) from Amplify to API 
Gateway. That means the REST API calls have to pass through the public internet to 
the API Gateway, which may induce some security compliance issue.

Hope the Amplify can enhance to support VPC integration, private domain name as well as WAF and ingress control.
Thank you.

**Comment by slikk66** [comment]
Also found this looking for a way to add WafV2 WebACL against the cloudfront in use by Amplify.  I suppose we'll have to generate our own ACM, Cloudfront distro and point to the amplify domain as origin, which decreases the helpfulness of Amplify considerably.  Would love to see it added.

**Comment by vishwas-mr** [comment]
@swaminator, any ETA as to when this feature would be available to users (specifically in ap-south-1)?

**Comment by yhminbv** [comment]
I need this feature!!!!!!

**Comment by apatel-rms** [comment]
IP Whitelisting or WAF would be very useful for me right now.

**Comment by BitShepherd** [comment]
We would also be interested in adding WAF to an Amplify App at my company.

**Comment by jfreeley-ninthwave** [comment]
+1 here too .. either Whitelisting or full feature waf ..

**Comment by grayaii** [comment]
We just got bit by this too.
We had an Amplify project that had custom domains.  We manually created a Cloudfront resource protected by WAF and manually tweaked Route53 to change the A Record from Amplify's internal Cloudfront distro to ours, and it didn't work.

It turns out you get really weird behavior if you do this. For instance, 50% of the time your domain goes to your CDN, the other times it goes to Amplify's endpoint.  DNS is not happy at all.

You MUST remove the management of the domain from your Amplify project.  BUT, like someone pointed out, the Amplify endpoint is public and if people know about it, they can circumvent your CDN all together.  But it's better than nothing!

**Comment by shetshrikant** [comment]
Any update on this one?

**Comment by abhijeet-toptal** [comment]
Any updates or ETA?

**Comment by rosmu** [comment]
This FR is pending for about 3 years now. We have put people on the moon with much less time than this. We expect to have this feature released before our next re:Invent

**Comment by emrul** [comment]
> This FR is pending for about 3 years now. We have put people on the moon with much less time than this. We expect to have this feature released before our next re:Invent

Not being funny, but when in history have people been put on the moon in less than 3 years?

**Comment by continueai-idan** [comment]
will this ever happen ?
using amplify without the ability to restrict access by source ip/deploying it internally  is not so good for internal apps .

**Comment by brian-snyder** [comment]
WAF with Amplify apps is highly desired/required feature for me as well.  Not just IP whitelisting but full feature set.

**Comment by royax88** [comment]
Any Updates?

**Comment by MaxHXie** [comment]
Chipping in on this also

**Comment by dgokcin** [comment]
any updates?

**Comment by justinwiley** [comment]
@garyleefight or maybe @swaminator any updates / can you comment on the priority of this feature in the roadmap?

**Comment by danfreid** [comment]
+1

**Comment by rodrigomotta1** [comment]
Does anyone knows any possible workaround for this integration between Amplify Hosting and WAF?

**Comment by omarcardona16** [comment]
Any Updates?

**Comment by marco910** [comment]
> Does anyone know any possible workaround for this integration between Amplify Hosting and WAF?

@rodrigomotta1 Maybe it would work to "wrap" the Amplify distribution with another CloudFront distribution and apply the WAF to that. I didn't test that and don't know how this would impact the performance, but maybe it works.

**Comment by Vlaaaaaaad** [comment]
> Does anyone know any possible workaround for this integration between Amplify Hosting and WAF?

It's possible through manual work according to [this blog post](https://www.xtivia.com/blog/aws-amplify-contentful-nextjs-tutorial/). I haven't tried it tho!

**Comment by rodrigomotta1** [comment]
@marco910 and @Vlaaaaaaad, thanks for the input! Unfortunately I decided to abandon Amplify Hosting and now I'm manually setting all up using CodePipeline, CodeBuild, CloudFront and WAF 😕 
For those having the same problem, I think that doing this job manually is a great workaround because you'll have a lot more control of what is happening behind the scenes, but will require some experience with AWS (CloudFormation can help a lot in this manual process also). [This reference](https://medium.com/@jeffreyrussom/react-continuous-deployments-with-aws-codepipeline-f5034129ff0e) helped me with the job.

Just for you to know, I'm setting up a CI/CD environment for a CRA React app, so if you're having issues like me, maybe this could work for you too! :)

**Comment by slikk66** [comment]
I posted about this in may '21 also..

I used Pulumi to setup my own cognito, cloudfront, s3 hosting, waf, appsync, dynamo, iam, lambdas, apigateway etc.  I still use the amplify front end libraries, but had to ditch their infrastructure side 100%

**Comment by marco910** [comment]
@Vlaaaaaaad Thanks for that approach! Unfortunately, this is not working anymore when using the new platform for Next.js 13. The resources for those distributions are completely hidden in the AWS console.

@rodrigomotta1 Sounds really great, but unfortunately, I think this is only working for static sites like yours which is created with Create-React-App. I'm working most of the with Next.js apps which are hosted serverless because of their SSR functions.
Do you know an approach, how this could work for Next.js apps?

@slikk66 So, are you still hosting on AWS or switched completely? What are you using now, if you switched?

**Comment by slikk66** [comment]
@marco910 still on AWS, but had lots of issues with Amplify infrastructure side. Just had too many bugs and issues that I ran into, this being one of them. I would deploy simple things with it and then see how they were configured, then copy that configuration to my own IaC tooling. So far has been working well.

**Comment by doguhanciftci** [comment]
Any updates?

**Comment by marco910** [comment]
@doguhanciftci No from AWS-side, but I built a CI/CD workflow on my own now.
Maybe you're interested in that setup: I'm using CodeBuild to build my Next.js site and then output the artifacts (build output) into a S3 bucket. After the build, a webhook on the VPS in triggered, which pulls the artifacts via AWS CLI from the bucket and deploys them locally on the instance. On top of the VPS, I'm running a CloudFront distribution with an additional WAF v2.
This is a bit of setup work, but in the end it works very well.

**Comment by doguhanciftci** [comment]
> @doguhanciftci No from AWS-side, but I built a CI/CD workflow on my own now. Maybe you're interested in that setup: I'm using CodeBuild to build my Next.js site and then output the artifacts (build output) into a S3 bucket. After the build, a webhook on the VPS in triggered, which pulls the artifacts via AWS CLI from the bucket and deploys them locally on the instance. On top of the VPS, I'm running a CloudFront distribution with an additional WAF v2. This is a bit of setup work, but in the end it works very well.

Thanks @marco910, since there is no hope on the Amplify side, I will implement a similar approach. Thanks for sharing your setup.

**Comment by amardesai1** [comment]
Any ETA for WAF or at least the ability to access the CloudFront Config for Amplify?

**Comment by costerman** [comment]
@swaminator - Loved the options you laid out a few years back in this issue thread (https://github.com/aws-amplify/amplify-hosting/issues/36#issuecomment-550449265)

Any update on option #2.  We are leveraging Amplify hosting for our Next JS React app and I need to do some specific IP range blocking, geographic restriction, and pattern matching (but in my case blacklisting instead of whitelisting unfortunately).

**Comment by hisham** [comment]
+1. Blacklisting ip required at minimum. Something like Cloudlflare bot management that just works out of the box with 1 click would be ideal for us who don't want to wrangle with AWS too much. Thanks!

**Comment by salauddinn** [comment]
has anyone tried this approch https://docs.aws.amazon.com/prescriptive-guidance/latest/patterns/enable-aws-waf-for-web-applications-hosted-by-aws-amplify.html

**Comment by costerman** [comment]
@salauddinn I haven't, but thank you for sharing this.  Will take a look and give it a try.

**Comment by srgustafson8** [comment]
> has anyone tried this approch https://docs.aws.amazon.com/prescriptive-guidance/latest/patterns/enable-aws-waf-for-web-applications-hosted-by-aws-amplify.html

I have, but wrote it using terraform and slightly less complicated (no caching, no lambda for password retrieval). It does work provided you get the credentials correct!

**Comment by salauddinn** [comment]
@srgustafson8 I'm planning to implement in Terraform. if possible can share the repo that you implemented? it might helpful for me to implement

**Comment by srgustafson8** [comment]
@salauddinn can't share the repo yet, we're not quite set up to open source anything at this point. 

tl;dr for the auth: 

using the random provider to generate a random username and password, with a variable that lets you rotate the credentials:
```hcl
resource "random_password" "username" {
...
  keepers          = { version = var.rotate_basic_creds }
}

resource "random_password" "password" {
...
  keepers          = { version = var.rotate_basic_creds }
}
```

then creating the base64 encoded result in a local var
```hcl
locals {
  creds = base64encode("${random_password.username.result}:${random_password.password.result}") }
}
```

add these as the values for basic auth in amplify and custom header in cloudfront
```hcl
resource "aws_cloudfront_distribution" "this" {
  web_acl_id = ...
...
  origin {
...
    custom_header {
      name  = "Authorization"
      value = sensitive("Basic ${local.creds}")
    }
  }
}

resource "aws_amplify_branch" "this" {
...
  enable_basic_auth      = true
  basic_auth_credentials = local.creds
}
```

And I just disabled caching on cloudfront entirely by adding the CachingDisabled cache policy. You could use a null resource doing a local exec to do the cache invalidation as the last step of the deployment pipeline which removes the need for a lambda to do it. 

Hope this helps!

**Comment by salauddinn** [comment]
i will try it Thank you @srgustafson8

**Comment by swaminator** [comment]
@salauddinn @srgustafson8 @hisham @costerman @amardesai1 @doguhanciftci @slikk66 @justinwiley @tchalvak @emrul @blazinaj @marco910 @rodrigomotta1 @Vlaaaaaaad @royax88 @danfreid @brian-snyder @denonade 

Amplify PM here - We are evaluating an approach that would connect a WAF resource in your account to the Amplify managed distribution. You will be able to use our console/API/CloudFormation to associate a WAF ACL (which you have configured in your account) with your Amplify app. However this approach has some near-term constraints around caching content and we want feedback if these constraints are acceptable:
1. Performance mode will not work
2. `Cache-control` headers set in custom headers would not be respected
3. Next.js ISR will not work
4. Next.js [cache control](https://nextjs.org/docs/pages/api-reference/next-config-js/headers#cache-control) settings would not work

Would love feedback on whether these constraints are acceptable as we decide whether we prioritize this project right now.

**Comment by amardesai1** [comment]
> @salauddinn @srgustafson8 @hisham @costerman @amardesai1 @doguhanciftci @slikk66 @justinwiley @tchalvak @emrul @blazinaj @marco910 @rodrigomotta1 @Vlaaaaaaad @royax88 @danfreid @brian-snyder @denonade 
> 
> Amplify PM here - We are evaluating an approach that would connect a WAF resource in your account to the Amplify managed distribution. You will be able to use our console/API/CloudFormation to associate a WAF ACL (which you have configured in your account) with your Amplify app. However this approach has some near-term constraints around caching content and we want feedback if these constraints are acceptable:
> 1. Performance mode will not work
> 2. `Cache-control` headers set in custom headers would not be respected
> 3. Next.js ISR will not work
> 4. Next.js [cache control](https://nextjs.org/docs/pages/api-reference/next-config-js/headers#cache-control) settings would not work
> 
> Would love feedback on whether these constraints are acceptable as we decide whether we prioritize this project right now. 

My company's amplify application is a Vite SPA , that the default caching behaviour will work fine with.

We'd love to see this functionality, regardless of these constraints. Looking forward to hearing more.

**Comment by slikk66** [comment]
> @salauddinn @srgustafson8 @hisham @costerman @amardesai1 @doguhanciftci @slikk66 @justinwiley @tchalvak @emrul @blazinaj @marco910 @rodrigomotta1 @Vlaaaaaaad @royax88 @danfreid @brian-snyder @denonade
> 
> Amplify PM here - We are evaluating an approach that would connect a WAF resource in your account to the Amplify managed distribution. You will be able to use our console/API/CloudFormation to associate a WAF ACL (which you have configured in your account) with your Amplify app. However this approach has some near-term constraints around caching content and we want feedback if these constraints are acceptable:
> 
> 1. Performance mode will not work
> 2. `Cache-control` headers set in custom headers would not be respected
> 3. Next.js ISR will not work
> 4. Next.js [cache control](https://nextjs.org/docs/pages/api-reference/next-config-js/headers#cache-control) settings would not work
> 
> Would love feedback on whether these constraints are acceptable as we decide whether we prioritize this project right now.

I'm not sure which "performance mode" you're referring to - but for the rest this would have worked for me.  I've since deployed my own WAF/distros/backend with Pulumi and use the front end Amplify libraries only to connect to my infra, but none of these constraints seem like deal breakers to me.  I think missing WAF ability for Amplify is a much bigger deal.. Amplify is meant to be used to build websites, yet lacks the ability to use native AWS tools to protect said websites.

**Comment by bsnyder74** [comment]
@swaminator Constraints seem reasonable to me.  +1 for prioritizing this please!!

**Comment by Vlaaaaaaad** [comment]
> Would love feedback on whether these constraints are acceptable as we decide whether we prioritize this project right now

@swaminator: yes, those are acceptable compromises for a faster launch. Launching something a bit more limited **today** and then iterating and eliminating the limitations is a good plan is this case!

That said, Amplify Hosting should have a default WAF included with the service. In my opinion, Amplify Hosting is a managed high-level service which means it should come with all batteries included (so with AWS Shield and AWS WAF configured and setup) but it should also support straightforward escape hatches (so allowing folks to use their own AWS WAF if they have custom or fancy requirements).

**Comment by frattaro** [comment]
@swaminator we have a public-facing app that it wouldn't be good for, a non-public one that it'd be beneficial to have that on. As long as it's configurable per-project there'd be interest

**Comment by srgustafson8** [comment]
> Would love feedback on whether these constraints are acceptable as we decide whether we prioritize this project right now.

Thanks @swaminator for the reply and great to hear you're looking in to this. The only dealbreaker there is the Next.JS ISR for us, which is key for our implementation. However, as long as that's going to be resolved in the future, it sounds like a positive step forward and our workaround will hold until the full implementation is ready.

**Comment by tchalvak** [comment]
Two questions from the approach, @slikk66

“
- [Performance mode](https://docs.aws.amazon.com/amplify/latest/userguide/ttl.html#Performance-mode) will not work
- Cache-control headers set in custom headers would not be respected
- Next.js ISR will not work
- Next.js [cache control](https://nextjs.org/docs/pages/api-reference/next-config-js/headers#cache-control) settings would not work
“
First, I assume that the standard index.html “revalidate” is not handled via cache-control headers, so the approach above allows for the standard use case?
Second, I think a common use case is creating alternate “dev” sites from “production” sites, for which the current basic auth approach - while feasible - is trouble. Will the new approach allow for the full power of WAF to create whitelisting of devs based on different criteria?

**Comment by thornyweb** [comment]
@swaminator was a decision made on this project progressing yet?

**Comment by swaminator** [comment]
Thank you everyone for your input. Given the feedback we got on the tradeoffs, we have unfortunately decided not to prioritize this feature this year. We understand that this may be disappointing to read, but we are going to spend some more time to deliver this feature right. We have prioritized two other top feature requests for delivery this year instead - #44 and #56.

**Comment by thornyweb** [comment]
> Thank you everyone for your input. Given the feedback we got on the tradeoffs, we have unfortunately decided not to prioritize this feature this year. We understand that this may be disappointing to read, but we are going to spend some more time to deliver this feature right. We have prioritized two other top feature requests for delivery this year instead - #44 and #56.  Thank you for your patience.

Hi @swaminator thank you for the update and your transparency. Are you able to confirm when this feature will be delivered?

Based on this update we need to make a business decision about whether or not to continue using Amplify.

Thanks.

**Comment by srgustafson8** [comment]
@swaminator This is quite disappointing - both of those requests combined have less 'thumbs' than this single ticket has and provide 'nice to have' functionality. Being able to protect public web endpoints is basic security practice and should be a day 1 requirement on every service that can be public facing, especially as the cyber threat landscape continues to become more dangerous.

**Comment by marco910** [comment]
@thornyweb We often face the same situation. Can I ask what alternatives to Amplify you are considering?

**Comment by thornyweb** [comment]
> @thornyweb We often face the same situation. Can I ask what alternatives to Amplify you are considering?

Within AWS it would be something containerized or just simple Node on an EC2 instance in a VPC with much better network configuration controls.

Extreme solution would be off of AWS entirely with a different provider.

**Comment by frattaro** [comment]
If a WAF was required for us, the first thing I'd do is look at cloudflare: https://www.cloudflare.com/application-services/products/waf/

Pretty sure it relies on DNS redirection and it never seemed all that secure to me but I'm sure there'd be ways to restrict usage to the WAF channel.

**Comment by mswezey23** [comment]
+1 Would love to have this feature developed as we roll out our merchant-side product to all merchants.

**Comment by afern247** [comment]
This is a must have requirement, I have my app now getting attacked on the main public url while I have it fully protected using Cloudflare.

+1 for this feature, very needed, critical.

**Comment by joon623** [comment]
Still no plans to use vpc, ip-whitelist, waf?

**Comment by milind-daftari** [comment]
Is there a temporary fix or workaround for this for now?

**Comment by srgustafson8** [comment]
> has anyone tried this approch https://docs.aws.amazon.com/prescriptive-guidance/latest/patterns/enable-aws-waf-for-web-applications-hosted-by-aws-amplify.html

@milind-daftari I followed this one that salauddin posted, it seems to work ok but not ideal

**Comment by em1l1000** [comment]
> > has anyone tried this approch https://docs.aws.amazon.com/prescriptive-guidance/latest/patterns/enable-aws-waf-for-web-applications-hosted-by-aws-amplify.html
> 
> @milind-daftari I followed this one that salauddin posted, it seems to work ok but not ideal

Would google still be able to crawl your website if you only allow one country in the allowed list?

**Comment by srgustafson8** [comment]
> > > has anyone tried this approch https://docs.aws.amazon.com/prescriptive-guidance/latest/patterns/enable-aws-waf-for-web-applications-hosted-by-aws-amplify.html
> > 
> > 
> > @milind-daftari I followed this one that salauddin posted, it seems to work ok but not ideal
> 
> Would google still be able to crawl your website if you only allow one country in the allowed list?

Probably not the right place to ask this question, but that being said, unless that one country is USA I'd imagine no.

**Comment by jimmyn** [comment]
Is there any update on this? This is very much needed feature!

**Comment by sawravchy** [comment]
Any update on this issue? In terms of security, this is a very much needed feature.

**Comment by frattaro** [comment]
I'd check back in September, because the last update we got was "we decided not to prioritize this feature this year" -- so I imagine they pick features on an annual basis.

**Comment by mathewmcnaughtonmass** [comment]
This is an important feature, i hope we can get its priority increased

**Comment by tuxillo** [comment]
Any news on the WAF support for Amplify?

**Comment by ppuvan** [comment]
Hello any update on this? We are planning to use Amplify for a large scale application in the medical sector, but not having a solid WAF/VPC implementation for Amplify are making our decision makers turn away from Amplify. Currently we are using the CloudFront routing approach by using Auth header and making Amplify private, but as pointed out above if the URL were to be compromised (as it is extremely likely), then we are opening ourselves to DoS attacks. Please prioritize this above all other Code First updates, as it should be simple for the likes of your kind to implement, considering it is a critical security improvement.

**Comment by LarsPede** [comment]
Echo'ing all that is being said here about the critically of the feature. We are looking into doing the cloudfront routing approach but as @ppuvan is saying - the fact that the default url is always public is a major security concern for us... Since it sounds like you aren't prioritizing adding the full blown feature - is there then a middleway where we can make the url not publicly available by default? it would allow us to use the cloudfront work-around and still leave us somewhat protected by the fact that only something like a specific policy has access to the url?

Definitely keep up the good work in general! Amplify is an amazing tool for getting up and running! Just sad to see that we are still in a "non-prod" state in terms of security 😞

**Comment by mauerbac** [comment]
@LarsPede @ppuvan @tuxillo @mathewmcnaughtonmass @jimmyn @joon623 
👋  Hello everyone -- we are in active development of this feature! Sorry for the lack of communication.

We are working through a few implementation decisions, and I'm curious to get a bit of feedback. We are exploring two approaches 1) An integration with the AWS WAF service 2) A more managed approach where you can easily accomplish IP allowlists, country  enable Bot Protection/DDoS Protection. 

Can you provide feedback on which approach would better suit your needs? I'm also available for a call if you can DM me on Twitter/X (@mauerbac) or email mauerbac@amazon.com to setup

**Comment by matt-dalton** [comment]
1 would definitely be better.

Would 2 allow us to set rules as we normally would with WAF? e.g. show a captcha based on request conditions?

**Comment by jimmyn** [comment]
@mauerbac definitely 1

**Comment by TheGreenInsurer-MT** [comment]
> @LarsPede @ppuvan @tuxillo @mathewmcnaughtonmass @jimmyn @joon623 👋 Hello everyone -- we are in active development of this feature! Sorry for the lack of communication.
> 
> We are working through a few implementation decisions, and I'm curious to get a bit of feedback. We are exploring two approaches 1) An integration with the AWS WAF service 2) A more managed approach where you can easily accomplish IP allowlists, country enable Bot Protection/DDoS Protection.
> 
> Can you provide feedback on which approach would better suit your needs? I'm also available for a call if you can DM me on Twitter/X (@mauerbac) or email [mauerbac@amazon.com](mailto:mauerbac@amazon.com) to setup

Either would work for me but 1 feels more complete

**Comment by sawravchy** [comment]
I would definitely prefer option 1.

**Comment by mauerbac** [comment]
@matt-dalton 
For #2, I'd be managed so you'd have fewer controls around request conditions and let Amplify Hosting optimize for DDoS protection against Layer 3, 4, 7 attacks. ie. If we suspect your site is under heavy attack, we'd force users into CAPTCHA challenge. The tradeoff is you don't have to worry/learn about WAF etc.. 

@jimmyn @sawravchy 
Makes sense! Do you already use WAF today? What features in WAF do you plan to leverage? 

Thank you for the super fast responses :)

**Comment by matt-dalton** [comment]
```I'd be managed so you'd have fewer controls around request conditions and let Amplify Hosting optimize for DDoS protection against Layer 3, 4, 7 attacks. ie. If we suspect your site is under heavy attack, we'd force users into CAPTCHA challenge. The tradeoff is you don't have to worry/learn about WAF etc..```
I mean that sounds nice in principle. My worry would be we'd be stuck if it proved ineffective...or we had specific requirements somehow.

**Comment by thornyweb** [comment]
> @matt-dalton For #2, I'd be managed so you'd have fewer controls around request conditions and let Amplify Hosting optimize for DDoS protection against Layer 3, 4, 7 attacks. ie. If we suspect your site is under heavy attack, we'd force users into CAPTCHA challenge. The tradeoff is you don't have to worry/learn about WAF etc..

Based on this reply, option 2 would be sufficient for my needs, while option 1 would be more complete.

I appreciate that you are still at a very early stage, but is there a significant time difference between delivering the 2 solutions? If option 1 was going to add months of extra development beyond number 2, I'd strongly advocate for the quicker delivery.

**Comment by jimmyn** [comment]
> @matt-dalton 
> 
> For #2, I'd be managed so you'd have fewer controls around request conditions and let Amplify Hosting optimize for DDoS protection against Layer 3, 4, 7 attacks. ie. If we suspect your site is under heavy attack, we'd force users into CAPTCHA challenge. The tradeoff is you don't have to worry/learn about WAF etc.. 
> 
> 
> 
> @jimmyn @sawravchy 
> 
> Makes sense! Do you already use WAF today? What features in WAF do you plan to leverage? 
> 
> 
> 
> Thank you for the super fast responses :) 

Yes we already use WAF for api gateway and appsync with custom rule sets, and it is configured in cloudformation, so for me it will be easier to be able to assign the same instance or create a separate one in cloudformation rather than a managed solution through the UI. But I guess both options should be implemented to cover all use cases.

**Comment by frattaro** [comment]
Option 2 because it's Amplify. What about new feature to "eject" from Amplify, where it provides you a cloud formation script?

**Comment by bsnyder74** [comment]
Whichever can be done the fastest and will actually get completed.  We've been waiting on this for years.
Can you provide any insights as to whether one will take longer than the other?

**Comment by mauerbac** [comment]
> > @matt-dalton For #2, I'd be managed so you'd have fewer controls around request conditions and let Amplify Hosting optimize for DDoS protection against Layer 3, 4, 7 attacks. ie. If we suspect your site is under heavy attack, we'd force users into CAPTCHA challenge. The tradeoff is you don't have to worry/learn about WAF etc..
> 
> Based on this reply, option 2 would be sufficient for my needs, while option 1 would be more complete.
> 
> I appreciate that you are still at a very early stage, but is there a significant time difference between delivering the 2 solutions? If option 1 was going to add months of extra development beyond number 2, I'd strongly advocate for the quicker delivery.

Gotcha -- my question is when you say option 1 is more complete -- would you mind doing the added work to learn and properly configure WAF? Is that something you'd invest the time to do? Not sure if you've checked out AWS WAF before

**Comment by thornyweb** [comment]
> > > @matt-dalton For #2, I'd be managed so you'd have fewer controls around request conditions and let Amplify Hosting optimize for DDoS protection against Layer 3, 4, 7 attacks. ie. If we suspect your site is under heavy attack, we'd force users into CAPTCHA challenge. The tradeoff is you don't have to worry/learn about WAF etc..
> > 
> > 
> > Based on this reply, option 2 would be sufficient for my needs, while option 1 would be more complete.
> > I appreciate that you are still at a very early stage, but is there a significant time difference between delivering the 2 solutions? If option 1 was going to add months of extra development beyond number 2, I'd strongly advocate for the quicker delivery.
> 
> Gotcha -- my question is when you say option 1 is more complete -- would you mind doing the added work to learn and properly configure WAF? Is that something you'd invest the time to do? Not sure if you've checked out AWS WAF before

@mauerbac  - I would if that was what was required, but as I say my preference is for the speediest solution, I and many others have been working with risk exceptions because of this for too long now.

There is an argument to be made for option 2 and a solution being integrated directly into Amplify.

Would the IP coverage in option 2 include allowing IP ranges and CIDR blocks as well as single fixed IP's? Would it cover both V4 and V6? Would it be both safe list and block list? Say I want to block all traffic except a handful of IP's in one application and in another, I want to allow all but block a select few.

**Comment by mauerbac** [comment]
> > > > @matt-dalton For #2, I'd be managed so you'd have fewer controls around request conditions and let Amplify Hosting optimize for DDoS protection against Layer 3, 4, 7 attacks. ie. If we suspect your site is under heavy attack, we'd force users into CAPTCHA challenge. The tradeoff is you don't have to worry/learn about WAF etc..
> > > 
> > > 
> > > Based on this reply, option 2 would be sufficient for my needs, while option 1 would be more complete.
> > > I appreciate that you are still at a very early stage, but is there a significant time difference between delivering the 2 solutions? If option 1 was going to add months of extra development beyond number 2, I'd strongly advocate for the quicker delivery.
> > 
> > 
> > Gotcha -- my question is when you say option 1 is more complete -- would you mind doing the added work to learn and properly configure WAF? Is that something you'd invest the time to do? Not sure if you've checked out AWS WAF before
> 
> @mauerbac - I would if that was what was required, but as I say my preference is for the speediest solution, I and many others have been working with risk exceptions because of this for too long now.
> 
> There is an argument to be made for option 2 and a solution being integrated directly into Amplify.
> 
> Would the IP coverage in option 2 include allowing IP ranges and CIDR blocks as well as single fixed IP's? Would it cover both V4 and V6? Would it be both safe list and block list? Say I want to block all traffic except a handful of IP's in one application and in another, I want to allow all but block a select few.

We are thinking the feature set would look like...

✅  Allow CIDR blocks and single fixed IPs
✅  IPv4 and IPv6
✅ : Allowlist
❌  Block list

**Comment by matt-dalton** [comment]
Not having a block list would have been limiting for us in the past, but obviously I've no idea how effective the #2 automated protections you're planning are.

**Comment by rwbayer** [comment]
As another data point, we'd prefer the first option, but agreed with speed being the primary driver here (would prefer the second option if it's significantly faster)

**Comment by sumitsahoo** [comment]
> @LarsPede @ppuvan @tuxillo @mathewmcnaughtonmass @jimmyn @joon623 👋 Hello everyone -- we are in active development of this feature! Sorry for the lack of communication.
> 
> We are working through a few implementation decisions, and I'm curious to get a bit of feedback. We are exploring two approaches 1) An integration with the AWS WAF service 2) A more managed approach where you can easily accomplish IP allowlists, country block, enable Bot Protection/DDoS Protection.
> 
> Can you provide feedback on which approach would better suit your needs? I'm also available for a call if you can DM me on Twitter/X (@mauerbac) or email [mauerbac@amazon.com](mailto:mauerbac@amazon.com) to setup

#1 looking more feasible. 

Also in our org, we use Imperva as WAF for all the applications. So flexibility needs to be there if a customer wants to implement a WAF other than AWS. We are trying to develop an application with Amplify for our org and sadly there are some org-level limitations that we need to follow (i.e. WAF).

**Comment by frattaro** [comment]
I'm surprised to see interest in Amplify by customers who require customized infra

**Comment by sumitsahoo** [comment]
> I'm surprised to see interest in Amplify by customers who require customized infra

Well, we are currently using Amplify Gen 1 but plan to migrate to Gen 2. It makes the development a bit faster instead of going with traditional 3-tier architecture. Also, GraphQL helps in many ways. At least in our org, we are now more committed to serverless components :) Less focus on infra and more focus on data :)

**Comment by LarsPede** [comment]
> I'm surprised to see interest in Amplify by customers who require customized infra
The fact that we don't have to worry about the rest of the infrastructure (fargate, ecs, what-not) makes it so easy to get up and running, that it outweighs the restrictions we currently have in amplify.
Automatic PR builds, automatic teardowns, automatic image optimization, the list goes on...

>  👋 Hello everyone -- we are in active development of this feature! Sorry for the lack of communication.
> 
> We are working through a few implementation decisions, and I'm curious to get a bit of feedback. We are exploring two approaches 1) An integration with the AWS WAF service 2) A more managed approach where you can easily accomplish IP allowlists, country block, enable Bot Protection/DDoS Protection.
> 
> Can you provide feedback on which approach would better suit your needs? I'm also available for a call if you can DM me on Twitter/X (@mauerbac) or email [mauerbac@amazon.com](mailto:mauerbac@amazon.com) to setup

@mauerbac Nice to hear!

Both would honestly be fine for us. Approach 1 would be preferred, since it allows us more control. My ultimate request would be the ability to opt into one or the other. I think ejecting fully isn't a valuable feature - if we want *full* control, then we wouldn't have started with amplify. I'm much more into opt'ing for the extra features that are required for prod runs of an application infrastructure.

**Comment by drew-carter** [comment]
> @LarsPede @ppuvan @tuxillo @mathewmcnaughtonmass @jimmyn @joon623 👋 Hello everyone -- we are in active development of this feature! Sorry for the lack of communication.
> 
> We are working through a few implementation decisions, and I'm curious to get a bit of feedback. We are exploring two approaches 1) An integration with the AWS WAF service 2) A more managed approach where you can easily accomplish IP allowlists, country block, enable Bot Protection/DDoS Protection.
> 
> Can you provide feedback on which approach would better suit your needs? I'm also available for a call if you can DM me on Twitter/X (@mauerbac) or email [mauerbac@amazon.com](mailto:mauerbac@amazon.com) to setup

Very keen on this implementation too. Option 1 would certainly be better for my use cases. Option 2 could suffice in many cases, but only if things like the IP lists were integrated with the WAFv2 IP Sets and potentially prefix lists.

**Comment by OlisaMarvis** [comment]
Wow, surprised this isn't available 1 year later.

**Comment by FavourOkonta** [comment]
Following up on this request? Any updates on WAF integration?

**Comment by MarlonJD** [comment]
Any updates? Can we use WAF with gen2 without any extra steps? What's the suggestion of Amplify team?

**Comment by magidandrew** [comment]
Seems like a poor man's solution is to programmatically create a lot of redirect rules with the associated country codes to perform geoblocking. I personally haven't tried this yet.

However, having direct access to cloudfront/WAF is much better

@swaminator Would love to help give feedback or do anything to accelerate this since this is a major blocking feature for our business use case.

Any updates on timelines would also be appreciated.

**Comment by mauerbac** [comment]
@OlisaMarvis @MarlonJD @FavourOkonta @magidandrew 

Hello everyone -- our Amplify + WAF integration is underway is currently being worked on. We are targeting November to launching this! I wish it were sooner but this is a realistic view of our timeline.

**Comment by LaurentDumont** [comment]
> @OlisaMarvis @MarlonJD @FavourOkonta @magidandrew
> 
> Hello everyone -- our Amplify + WAF integration is underway is currently being worked on. We are targeting November to launching this! I wish it were sooner but this is a realistic view of our timeline.

Would there be an open beta before November?

**Comment by arturkasperek** [comment]
@mauerbac how things look like - November is still real?

**Comment by mauerbac** [comment]
Hi everyone -- this is on track! We have alignment with the different AWS teams involved (WAF, CloudFront and Amplify). Will provide an update next month!

**Comment by ppuvan** [comment]
@mauerbac this is great news!! thanks for everything you guys are doing in Amplify and our team couldn't be more happier to hear this news.

**Comment by magidandrew** [comment]
Hey @mauerbac thanks for the update!! Could you let us know when you plan to make this announcement in September?  

We prefer option1 immensely over option 2, and would be great to understand what the teams have decided to implement

Again, super great news and really excited to hear this is coming underway. We are planning around _extremely_ tight business deadlines and knowing what's going to be in the scope is super critical for us

**Comment by arimanyus** [comment]
> Hi everyone -- this is on track! We have alignment with the different AWS teams involved (WAF, CloudFront and Amplify). Will provide an update next month!

Perfect! can't wait for it.

**Comment by rarias-mdp** [comment]
> Hi everyone -- this is on track! We have alignment with the different AWS teams involved (WAF, CloudFront and Amplify). Will provide an update next month!

Thanks, that will be useful!

**Comment by MyNameIsTakenOMG** [comment]
> Hi everyone -- this is on track! We have alignment with the different AWS teams involved (WAF, CloudFront and Amplify). Will provide an update next month!

This is GREAT!!!

**Comment by dropcontrol** [comment]
> @OlisaMarvis @MarlonJD @FavourOkonta @magidandrew
> 
> Hello everyone -- our Amplify + WAF integration is underway is currently being worked on. We are targeting November to launching this! I wish it were sooner but this is a realistic view of our timeline.

This is wonderful news!! However, I’m wondering if this feature will also be available for Gen1. There are still many Gen1 users, and since there is no migration tool, users who cannot move to Gen2 (myself included) are hoping it will be provided for Gen1 as well.

**Comment by msudol** [comment]
> > @OlisaMarvis @MarlonJD @FavourOkonta @magidandrew
> > Hello everyone -- our Amplify + WAF integration is underway is currently being worked on. We are targeting November to launching this! I wish it were sooner but this is a realistic view of our timeline.
> 
> This is wonderful news!! However, I’m wondering if this feature will also be available for Gen1. There are still many Gen1 users, and since there is no migration tool, users who cannot move to Gen2 (myself included) are hoping it will be provided for Gen1 as well.

I'd like to know the answer on this as well. Will the planning WAF integration work on Gen1?

**Comment by mauerbac** [comment]
Short answer - yes, WAF will work on Gen1 and Gen2 

hi @msudol & @dropcontrol - good question. Amplify Gen1 and Gen2 refer to the backend offering for Amplify. As it stands, Amplify Hosting is frontend hosting and doesn't have a notion of Gen1 and Gen2, so all customers will have this WAF feature :)

**Comment by LaurentDumont** [comment]
@mauerbac I'm reaching out to our AWS TAM for the official request, but any chance of an early beta/tech preview.

It would be cool to test-drive the feature before a full November release!

**Comment by MyNameIsTakenOMG** [comment]
> @mauerbac I'm reaching out to our AWS TAM for the official request, but any chance of an early beta/tech preview.
> 
> It would be cool to test-drive the feature before a full November release!

Agree! An early beta would be really useful. @mauerbac Sir, do you think whether we can have a beta version in early this month?

**Comment by SantoshRaoMea** [comment]
Will really help is the announcement on this is made please, desperately waiting for this one

**Comment by mswezey23** [comment]
Updates?

**Comment by mauerbac** [comment]
Hey all -- we are still on track for end of November! The _second_ we have a beta version/early access I will post here! Thank you for being patient 😄

**Comment by wgcv** [comment]
We are on November waiting any news! 🥳

**Comment by sebastianpinosanchez** [comment]
We are also waiting for this release. Continuity with Amplify depends on this feature.

**Comment by AfricaBlue72** [comment]
For our enterprise this feature is essential in order to consider AWS Amplify Hosting.

**Comment by mauerbac** [comment]
Hey everyone - sorry for the delay, this has been a day by day situation as we were hoping to have something out this week. Unfortunately, the testing period took longer than expected and with restrictions around AWS re:invent we will not be able to make our end of month goal. We will now plan to launch in mid-december, in just 2-3 weeks. The second something is ready it will be shared here. The team has been working on this feature with the upmost priority and urgency as we were really hopeful to have this out by re:invent. I know this is the most critical feature we build, so apologize for the delay in delivery here. 🙏

**Comment by sumitsahoo** [comment]
We kind off went the other way after seeing so much delay in `WAF` implementation. We removed `Amplify` hosting from the app and hosted our frontend in `AppRunner` using `ECR` `Docker` image. This gives more flexibility in integrating any `WAF` that we want. The reason for this move is because our use case is enterprise one and hence `WAF` is absolutely mandatory and in our case it has to be `Imperva WAF`.

**Comment by benrosen78** [comment]
When is this expected to launch now?

**Comment by sademban** [comment]
when do we expect this changes @mauerbac ?  i hit a roadblock of securing one of my micro frontend to a specific area as there was no way to do it apart from leaving it public.

**Comment by aha-kid** [comment]
Updated, I see it available as preview now in aws amplify console and you can choose to create a new WAF first then change some setting after created in "View WAF logs" or just use existed WAF rules. Thanks AWS Team

**Comment by benrosen78** [comment]
That is great to hear. What does it mean by being marked as a "Preview"?

**Comment by aha-kid** [comment]
Still in "beta" I guess. But still better to have it.

**Comment by TheGreenInsurer-MT** [comment]
Just had a quick first glance, it looks a very promising start, however I'm still not sure it's doing what I need so will have to spend time reviewing it properly.

It appears that settings are all across the entire app, e.g. IP restrictions aren't configurable by branch, so if you are using feature or preview branches you won't be able to restrict those without also restricting your main production branch. That means I may need to look at creating separate apps for production and non-prod environments so they can have different firewall configurations. That's not necessarily a bad thing but isn't how I was expecting it to work given that the Amplify apps support multiple branches and preview builds.

**Comment by thornyweb** [comment]
Just had a quick first glance, it looks a very promising start, however I'm still not sure it's doing what I need so will have to spend time reviewing it properly.

It appears that settings are all across the entire app, e.g. IP restrictions aren't configurable by branch, so if you are using feature or preview branches you won't be able to restrict those without also restricting your main production branch. That means I may need to look at creating separate apps for production and non-prod environments so they can have different firewall configurations. That's not necessarily a bad thing but isn't how I was expecting it to work given that the Amplify apps support multiple branches and preview builds.

**Comment by rotem1850** [comment]
Does someone know what this configuration does?
![image](https://github.com/user-attachments/assets/9f9264a9-8b88-4e91-af6c-304f95e60c58)

i can just turn it on or off

**Comment by aha-kid** [comment]
> Does someone know what this configuration does? ![image](https://private-user-images.githubusercontent.com/26322317/396846743-9f9264a9-8b88-4e91-af6c-304f95e60c58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ1MTQ5ODYsIm5iZiI6MTczNDUxNDY4NiwicGF0aCI6Ii8yNjMyMjMxNy8zOTY4NDY3NDMtOWY5MjY0YTktOGI4OC00ZTkxLWFmNmMtMzA0Zjk1ZTYwYzU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE4VDA5MzgwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0NjUyNmZiNGQyYmQzNTFlZDBkMmRmMGFmMjNiNzhiMTE5N2QxM2UyMGQ3YTM2ZTM5N2FmYzFiZWMzZjNlZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UZ00qCBm-_R5H1Mbn5o8XFGZnRRK9Z587ylTkkyGBZI)
> 
> i can just turn it on or off

I think it for who has changed domain to custom domain and doesn't want anyone try to access website by the amplify original website like "https://xxx.yyyy.amplifyapp.com"

For me, I turned it on so googlebot won't reach my website and marked it as duplicated

**Comment by swaminator** [comment]
Wow you guys are on top of everything :)
Seasons Greetings! We officially announce a preview release of our integration with WAF: https://docs.aws.amazon.com/amplify/latest/userguide/WAF-integration.html. This is a preview release so we get early feedback on usability - the feature itself is ready to use to secure your application. Thank you for your patience for this feature. It's finally here!

**Comment by benrosen78** [comment]
Thank you @swaminator !!

**Comment by ferdingler** [comment]
> It appears that settings are all across the entire app, e.g. IP restrictions aren't configurable by branch, so if you are using feature or preview branches you won't be able to restrict those without also restricting your main production branch. That means I may need to look at creating separate apps for production and non-prod environments so they can have different firewall configurations. That's not necessarily a bad thing but isn't how I was expecting it to work given that the Amplify apps support multiple branches and preview builds.

@thornyweb thanks for the feedback. You are right that the WebACL Configuration applies to the entire App, including all branches and all custom domains. As you mentioned, the natural solution is to split your Dev branch and your Prod branch into separate Apps. However, if you really want to keep them both on the same App, here is a little trick you can do to restrict IP addreses on your dev branch only. 

Say that your dev branch is accessed via https://dev.example.com and your prod branch is accessed by https://www.example.com. You can create an IP restriction rule that matches the `host` header so that it only applies to dev.example.com (aka your dev branch). Here is the rule represented as JSON. Notice the `NOT` negation is important because it makes the rule block the request when it **does not originate from an allowed IP** AND when the request is for dev.example.com. 

```json
{
  "Name": "IpAllowList-DevBranch",
  "Priority": 1,
  "Statement": {
    "AndStatement": {
      "Statements": [
        {
          "NotStatement": {  // <---- this is important.
            "Statement": {
              "IPSetReferenceStatement": {
                "ARN": "arn:aws:wafv2:us-east-1:111111111111:global/ipset/AmplifyIPSet/8c2cebcc6521"
              }
            }
          }
        },
        {
          "ByteMatchStatement": {
            "SearchString": "dev.example.com",
            "FieldToMatch": {
              "SingleHeader": {
                "Name": "host"
              }
            },
            "TextTransformations": [
              {
                "Priority": 0,
                "Type": "NONE"
              }
            ],
            "PositionalConstraint": "CONTAINS"
          }
        }
      ]
    }
  },
  "Action": {
    "Block": {}  // <---- this is important.
  }
}
```

and ensure that the default Action on your WebACL is set to Allow:

<img width="691" alt="Screenshot 2024-12-18 at 10 01 37 AM" src="https://github.com/user-attachments/assets/102c4c69-1c74-45d6-8bbc-39db206f95e1" />

I hope it helps!

**Comment by ferdingler** [comment]
> Still in "beta" I guess. But still better to have it.

To clarify, the feature is production ready. It isn't beta. So, feel free to use it in your production workloads with confidence. During the Preview period however, we want to gather feedback from customers and also iron out a few limitations that aren't available yet during this period, for example: 

- During the preview period, Amplify supports partial integration with CloudTrail. One management event during web ACL association will not appear in your CloudTrail logs.
- During preview, when your web ACL is associated to an Amplify resource, this new Amplify resource will not display in the Associated resources tab in the WAF console. You need to use the Amplify console to view the WebACL associated to an App. 
- During preview, AWS Config integration will not be available.
- During preview, the feature is not available in opt-in regions: ap-east-1, me-south-1 and eu-south-1.

**Comment by OlisaMarvis** [comment]
Just checking this out and I'm having tears in my eyes. AWS thank you,
amplify has won me over.😁

**Comment by hiroki-gota** [comment]
Same here! Just in time Xmas gift from AWS ❤️

**Comment by MarlonJD** [comment]
Hey there, I'm using only backend in amplify, not amplify hosting, Is this waf can protect my backend or I don't need this?

**Comment by MyNameIsTakenOMG** [comment]
> Hey there, I'm using only backend in amplify, not amplify hosting, Is this waf can protect my backend or I don't need this?

Well, I don't think this gonna work in your case, because this firewall feature is integrated with amplify hosting service, not with your amplify backend. However, you can still use WAF to protect your backend as WAF can integrate with resources like, api gateway if you have one

**Comment by MarlonJD** [comment]
Hey @MyNameIsTakenOMG Thanks for the answer, actually not API gateway but I'm using GraphQL API via AppSync, I don't know if AppSync need to be protect like in WAF, if it is, I'll think about this.

**Comment by blakedunson** [comment]
> Hey @MyNameIsTakenOMG Thanks for the answer, actually not API gateway but I'm using GraphQL API via AppSync, I don't know if AppSync need to be protect like in WAF, if it is, I'll think about this.

fwiw, [AppSync supports WAF integration](https://docs.aws.amazon.com/appsync/latest/devguide/WAF-Integration.html)

**Comment by sademban** [comment]
wow thank you AWS Amplify team for the early christmas gift, i was stuck in public traffic as it was not possible before. Thank you @mauerbac @ferdingler @swaminator <3. Let me test this and come back for the review 😄

**Comment by sademban** [comment]
hello @mauerbac @ferdingler @swaminator , it worked like a charm and thank you.

**issue**:
once i enabled the firewall, the error page that is being shown to the restricted ip is cloudfront error:
![image](https://github.com/user-attachments/assets/7d45029f-2068-4443-a827-b1a86b26c9a6)

1. I tried using redirect rule and tried to use a custome 403.html 
![image](https://github.com/user-attachments/assets/b43a71ab-ebfe-4550-b353-00fa978079a5)

Still i could not show my own custome error message page. 

is there anyway i could show custome error page on 403 so that cloudfront error page is not shown ??

HELP !!!

Thank You

**Comment by ferdingler** [comment]
@sademban Yes, adding a custom HTML response is easy, but it needs to be done directly in the WAF console. Here are the high level steps: 

1. Go to the WebACL that was created by Amplify and go to the Custom response bodies. 

<img width="1678" alt="Screenshot 2024-12-19 at 8 45 53 PM" src="https://github.com/user-attachments/assets/46d20a09-7873-40e6-a68f-38ceb95eda05" />

2. Create your custom response as HTML

<img width="1075" alt="Screenshot 2024-12-19 at 8 42 33 PM" src="https://github.com/user-attachments/assets/d4817833-6ef8-410d-b0ce-49e0527bf505" />

3. Then go the Rule that you want to apply the custom response to and click Edit. 
 
<img width="1670" alt="Screenshot 2024-12-19 at 9 03 37 PM" src="https://github.com/user-attachments/assets/3eafb91b-3431-49a9-b63c-5028795b7f7f" />

4. Select the custom response you just created in the `Action` section. 

<img width="824" alt="Screenshot 2024-12-19 at 8 43 37 PM" src="https://github.com/user-attachments/assets/88d92ac4-7bd3-4dde-899a-77a072b02fc0" />

---

That's it. You should now get your custom HTML. 

Here is the official documentation for custom response if you want to take a deeper look at the feature: https://docs.aws.amazon.com/waf/latest/developerguide/customizing-the-response-for-blocked-requests.html

**Comment by sademban** [comment]
Hello @ferdingler 
To meet my requirements, i tried to fllow but i was not successfule:

1. I created firewall settings for my amplify app
![image](https://github.com/user-attachments/assets/376e007b-4c17-42a3-b08c-6c24efe86487)

2. I fllowed your above instruction 

- created `custom response bodies` and saved it
![image](https://github.com/user-attachments/assets/44e0d4b5-2026-4be5-8d2f-0b2c9b7960d4)

![image](https://github.com/user-attachments/assets/2826bf74-19f8-4179-a79e-229bfed77ee2)

-  went to `Rules` and attached custome response to the `BlockAmplifyDomain`
![image](https://github.com/user-attachments/assets/f248c8de-4e82-4493-9036-b2003f9e7f8a)

![image](https://github.com/user-attachments/assets/4c1765a7-5a04-490f-844e-89b2efda85c6)

![image](https://github.com/user-attachments/assets/59ab298c-e2d7-43a0-aeb6-149218269b3f)

4. I also rebuilt the amplify but whe i try to acess from the ip which is not allowed in the amplify firewall i still see this
![image](https://github.com/user-attachments/assets/0205c107-a032-4f7f-8af7-0151ab314167)

will you able to cross check and help me.

thank you

**Comment by sademban** [comment]
@mauerbac @swaminator @ferdingler  any update ??, i could not show the custome HTML error page as  per the solution provided by @ferdingler . Is there mistake in my configuration or is it the support for firewall error page is not implmented?

**Comment by mauerbac** [comment]
hi @sademban - just to confirm. The rules are working correctly for you, right? Meaning when you access for a blocked IP the page is blocked correctly. The issue is you don't see the custom 404 page?

**Comment by goldbez** [comment]
@sademban, this is likely because you are accessing the amplifyapp domain from an IP address that is not allow-listed. The IpAllowList rule takes precedent over the BlockAmplifyDomain rule in your configuration. You do not have a custom response body associated with the IpAllowList rule. Please try associating the custom response body to the IpAllowList rule, too.

**Comment by sademban** [comment]
hello @mauerbac  yes the rules are working correctly , when we try to access from blocked ip we get the cloudfront default 403 page but i wanted to show a custom page. I followed @ferdingler 's steps.

hello @goldbez let me check because i dont think there will be a option to load customeHTML in custome response in IpAllowList  rule.

**Comment by goldbez** [comment]
@sademban The WebACL and all of its rules exist in your AWS account. You have the ability to edit them in any way that you would like. You can add the custom response body to the IpAllowList in the same way that you added it to the BlockAmplifyDomain rule.

**Comment by mauerbac** [comment]
Out of curiosity @sademban - did you create the WAF via the Amplify console? then go to WAF console to load the custom HTML?

**Comment by mauerbac** [comment]
📣 Feedback for the integration so far? How is it working? What could have been smoother?

**Comment by sademban** [comment]
hello @mauerbac integraion so far is working fine as expected apart from the cutsome error response page while accesing the site from ip which are not allowed. I haven't got time to test the custome HTML response for error page yet, I will follow the suggestion of yours and @goldbez  and give you a feedback on that as soon as i can test it. Thank you for the WAF intgration in Amplify it saved me from lot of trouble . 😄

**Comment by FavourOkonta** [comment]
Great job done to the AWS team here.
This comment is to the WAF team here concerning IPSets. 
With the use of IPv6; it appears IPSET automatically converts the address to the canonical form; making the actual/real address to be different. 

Example: A real IP of 2024:8a0:ec73:f800:988:28ee:d3a7:123 is converted on the WAF console to 2024:08a0:ec73:f800:0988:28ee:d3a7:0123 which when the rules are evaluated on WAF; automatically directs to the "Default Action" which may be "BLOCK" Action in specific case.

Is it possible for the WAF team to work on the non-normalization of these values; so same values put in are seen and evaluated by the console...

<img width="387" alt="image" src="https://github.com/user-attachments/assets/8e5aab8b-26a4-421a-a364-bb8db06fb501" />

<img width="527" alt="image" src="https://github.com/user-attachments/assets/a8d7a79e-74d6-43e9-86d8-cbd1039ab2bd" />

**Comment by anilkumaran** [comment]
Would love to have port-based allow/block (like block http)

**Comment by mauerbac** [comment]
Hey everyone -- I was on a leave from AWS, but now I'm back! 

Excited to see some positive comments around WAF being released. We are currently working towards our GA launch next month. I'm very curious to hear how the WAF integration has been working for folks? Even if it's working as expected, it's helpful to hear that! 

What would you like to see? What could be working better? Please let me know! Thank you!

**Comment by TheGreenInsurer-MT** [comment]
> Hey everyone -- I was on a leave from AWS, but now I'm back!
> 
> Excited to see some positive comments around WAF being released. We are currently working towards our GA launch next month. I'm very curious to hear how the WAF integration has been working for folks? Even if it's working as expected, it's helpful to hear that!
> 
> What would you like to see? What could be working better? Please let me know! Thank you!

Hi @mauerbac 

The WAF integration is working well for us. We implemented it on a couple of our sites very early after it was released and have recently rolled it out across all of our sites. It's working as expected.

We ended up using WAF properties beyond the capability of the Amplify console, but using the WAF console was easy enough and then very straightforward to link them up.

It was a bit of a pain having to split out and duplicate apps so we could have different rules for different branches of the same repo. Especially as there was no easy way to lift the environment variables in the console from another app so had to manually move each one for each app.

Obviously at the moment there is no pricing but I know that will change with the GA launch so I think it's important that there is a a fair amount of lead up time where the pricing has been announced in case people need to make changes if the cost doesn't work for them.

**Comment by mauerbac** [comment]
Thanks @TheGreenInsurer-MT for taking the time to write this! Good to hear and glad it was easy to go from Amplify created WAF to WAF console for full fledged features. 

Regarding pricing - I will provide an update here as soon as it's finalized. As of now, it is looking to be a small flat monthly per app fee. You would also incur charges to the WAF service. 

Curious if others have feedback

**Comment by lifeodyssey** [comment]
Hello @mauerbac Thank you for your work about this amazing integration. It works well. 
I'm trying to keep the versioning of waf rule using terraform. But it seems amplify currently do not support terraform to integrate with waf. The waf should be deployed to us-east-1 as it is for cloudfront. But my amplify is not in this area. This seems to be the problem. Do you have any idea about this?

**Comment by wlee221** [comment]
Hi @lifeodyssey, I'm from AWS Amplify. 

WAF webACL region does not need to match the Amplify region for the integration:
- webACL must be in `us-east-1`, as you mentioned, with `CLOUDFRONT` scope. This is also known as `"Global (CloudFront)"` in some interfaces. 
- Amplify app can be in any of the non opt-in regions that it supports. They don't have to match the webACL region. WAF / Amplify will handle the cross-region integration on behalf of you. 
    - Note that opt-in region support will be in the General Availability (GA) launch. 

For the terraform support, I recommend reaching out to the terraform [here](https://github.com/hashicorp/terraform-provider-aws/issues?q=is%3Aissue%20state%3Aopen%20amplify) for Amplify support. But have you tried connecting through [`aws_wafv2_web_acl_association`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/wafv2_web_acl_association) already? You should be able to define association through `wafv2` if you haven't tried so already.

**Comment by lifeodyssey** [comment]
> Hi [@lifeodyssey](https://github.com/lifeodyssey), I'm from AWS Amplify.
> 
> WAF webACL region does not need to match the Amplify region for the integration:
> 
> * webACL must be in `us-east-1`, as you mentioned, with `CLOUDFRONT` scope. This is also known as `"Global (CloudFront)"` in some interfaces.
> * Amplify app can be in any of the non opt-in regions that it supports. They don't have to match the webACL region. WAF / Amplify will handle the cross-region integration on behalf of you.
> 
> For the terraform support, I recommend reaching out to the terraform [here](https://github.com/hashicorp/terraform-provider-aws/issues?q=is%3Aissue%20state%3Aopen%20amplify) for Amplify support. But have you tried connecting through [`aws_wafv2_web_acl_association`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/wafv2_web_acl_association) already? You should be able to define association through `wafv2` if you haven't tried so already.

Hello @wlee221
Thank you for your reply. I have tried the `aws_wafv2_web_acl_association`.
in a way like this
```terraform
resource "aws_wafv2_web_acl_association" "amplify_waf" {
  resource_arn =  aws_amplify_app.app.arn
  web_acl_arn  = var.waf_arn
}
```
but it always show a error like
> creating WAFv2 WebACL Association (arn:aws:wafv2:us-east-1:xxxx:global/webacl/abcx/xxx,arn:aws:amplify:ap-southeast-2:xxxx:apps/xxxx): operation error WAFV2: AssociateWebACL, https response error StatusCode: 400, RequestID: xxx, WAFInvalidParameterException: Error reason: The ARN isn't valid. A valid ARN begins with arn: and includes other information separated by colons or slashes., field: RESOURCE_ARN, parameter: arn:aws:wafv2:us-east-1:xxxxx:global/webacl/astra-dev-waf/xxxx

I checked the code several times and could not found what may be the reason besides region. But the interesting thing is that I can associate them in the aws console.

**Comment by wlee221** [comment]
Hi @lifeodyssey,

that error usually happens if the wafv2 call is being made from a region that is not `us-east-1`. Terraform must call web acl association (`wafv2:AssociateWebACL`) in the `us-east-1` region, or the association will fail with that generic error message. This region is something that terraform might have to set explicitly for Amplify associations, if your terraform is not on `us-east-1` by default.

**Comment by lifeodyssey** [comment]
> Hi [@lifeodyssey](https://github.com/lifeodyssey),
> 
> that error usually happens if the wafv2 call is being made from a region that is not `us-east-1`. Terraform must call web acl association (`wafv2:AssociateWebACL`) in the `us-east-1` region, or the association will fail with that generic error message. This region is something that terraform might have to set explicitly for Amplify associations, if your terraform is not on `us-east-1` by default.

good it works! Thanks a lot !

**Comment by mauerbac** [comment]
Hey everyone! This Amplify Web Application Firewall has gone GA (General availability) today! 

You can read more here -> https://aws.amazon.com/blogs/aws/firewall-support-for-aws-amplify-hosted-sites

Please note, we have introduced a $15 /month per app charge for this feature. If you do not wish to pay that fee please make sure you dissociate the Web ACL from your Amplify app as this was free in Preview.
